File: packages/oxlint-plugin-react-doctor/src/plugin/rules/bundle-size/no-full-lodash-import.ts
import { defineRule } from "../../utils/define-rule.js";
import { isImportAbsentFromClientBundle } from "../../utils/is-import-absent-from-client-bundle.js";
import { isOutsideBrowserBundle } from "../../utils/is-outside-browser-bundle.js";
import { isTypeOnlyImport } from "../../utils/is-type-only-import.js";
import type { RuleContext } from "../../utils/rule-context.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";

export const noFullLodashImport = defineRule({
  id: "no-full-lodash-import",
  title: "Full lodash import",
  tags: ["test-noise"],
  severity: "warn",
  recommendation:
    "Import just the function you need: `import debounce from 'lodash/debounce'`. Saves about 70kb.",
  create: (context: RuleContext) => ({
    ImportDeclaration(node: EsTreeNodeOfType<"ImportDeclaration">) {
      const source = node.source?.value;
      // `lodash-es` ships ES modules that bundlers can tree-shake
      // (each function is a separate file); only the legacy bundled
      // `lodash` import pulls the whole library. Flagging
      // `lodash-es` would just push users to a more awkward import
      // form for the same byte cost.
      if (source !== "lodash") return;
      // Type-only imports are erased at emit time, so they ship nothing.
      if (isTypeOnlyImport(node)) return;
      // Bindings referenced only in type positions or Next.js server data
      // functions never reach the client bundle either.
      if (isImportAbsentFromClientBundle(node)) return;
      if (isOutsideBrowserBundle(node, context.filename)) return;
      context.report({
        node,
        message:
          "Importing all of lodash ships the whole library to your users & slows page load. Import from 'lodash/functionName' instead.",
      });
    },
  }),
});

File: native/oxlint/rules/no-full-lodash-import.rs
use oxc_ast::AstKind;
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;

use crate::{
    AstNode,
    context::{ContextHost, LintContext},
    rule::Rule,
};

const MESSAGE: &str = "Importing all of lodash ships the whole library to your users & slows page load. Import from 'lodash/functionName' instead.";

#[derive(Debug, Default, Clone)]
pub struct NoFullLodashImport;

declare_oxc_lint!(
    /// Disallow imports from the full legacy lodash bundle.
    NoFullLodashImport,
    react_doctor_native,
    perf,
    version = "0.1.0",
    short_description = "Disallow imports from the full legacy lodash bundle.",
);

impl Rule for NoFullLodashImport {
    fn should_run(&self, ctx: &ContextHost) -> bool {
        !is_non_production_file(ctx)
    }

    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::ImportDeclaration(declaration) = node.kind() else {
            return;
        };
        if declaration.source.value != "lodash"
            || declaration.import_kind.is_type()
            || is_import_absent_from_client_bundle(declaration, ctx)
            || is_outside_browser_bundle(ctx)
            || lodash_is_library_dev_page(ctx)
        {
            return;
        }
        ctx.diagnostic(OxcDiagnostic::warn(MESSAGE).with_label(declaration.span));
    }
}

fn lodash_is_library_dev_page(ctx: &LintContext<'_>) -> bool {
    let filename = ctx.file_path().to_string_lossy().replace('\\', "/");
    filename
        .rsplit('/')
        .next()
        .is_some_and(|basename| basename.contains(".page."))
        && is_published_library_package(ctx.file_path())
}
